### ISSUE TITLE
[BUG][v1.5.4-rc1] Recurring job failed to create/delete backups after node reboot

### ISSUE THREAD

**Issue by yangchiu** [longhorn/longhorn#7854]
## Describe the bug

1. Create a volume and create a backup recurring job with cron `* * * * *` to create a backup every minute and `retain = 1` to only keep one backup. 
2. Keep writing data to the volume.
3. Reboot a node of a replica. After the node and the volume are back to work. The recurring job could still work abnormally. Lots of `Error` backup remain to the backup list:
![backup-error](https://github.com/longhorn/longhorn/assets/5709208/83cb66d2-a3f0-4dbc-87f8-e9eff9593a18)
Even though a very long time after the reboot event, newly created backup could still randomly run into `Error` state with error message:
```
Error:                      proxyServer=10.42.3.10:8501 destination=10.42.1.72:10000: 
failed to backup snapshot vol-yde7-971b6006-6f18-4d12-85dc-c337a72d5220 to backup-5b793b248f96439d: 
rpc error: code = Unknown desc = failed to create backup to s3://backupbucket@us-east-1/backupstore 
for volume vol-yde769: rpc error: code = Unknown 
desc = failed lock backupstore/volumes/e2/eb/vol-yde769/locks/lock-5242291888d04df6.lck type 1 acquisition
```

## To Reproduce

Checkout PR https://github.com/longhorn/longhorn-tests/pull/1539 and run test case `Reboot Replica Node While Heavy Writing And Recurring Jobs Exist`

## Support bundle for troubleshooting

[supportbundle_359c9225-c5aa-42d3-9af0-6a301e68c081_2024-02-06T05-54-16Z.zip](https://github.com/longhorn/longhorn/files/14175471/supportbundle_359c9225-c5aa-42d3-9af0-6a301e68c081_2024-02-06T05-54-16Z.zip)

## Environment

 - Longhorn version: v1.5.4-rc1
 - Impacted volume (PV): 
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: v1.28.4+k3s1
   - Number of control plane nodes in the cluster: 1
   - Number of worker nodes in the cluster: 3
 - Node config
   - OS type and version: sles 15-sp5
   - Kernel version:
   - CPU per node:
   - Memory per node:
   - Disk type (e.g. SSD/NVMe/HDD):
   - Network bandwidth between the nodes (Gbps):
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal):
 - Number of Longhorn volumes in the cluster:

**Comment by innobead** [comment]
I assume this also happened in 1.6.x.

**Comment by innobead** [comment]
@james-munson Please help with this. Thanks.

**Comment by yangchiu** [comment]
Could be difficult to reproduce. The test case could fail due to different reasons instead of what described in https://github.com/longhorn/longhorn/issues/7854#issue-2120056364.

It seems hard to predict when will a backup creation can be triggered by a recurring job after a node reboot.

@james-munson Please don't try to reproduce it by running the test case since it could be diverse. The test case needed to be refined first. Please look into the support bundle or access the cluster by:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJkekNDQVIyZ0F3SUJBZ0lCQURBS0JnZ3Foa2pPUFFRREFqQWpNU0V3SHdZRFZRUUREQmhyTTNNdGMyVnkKZG1WeUxXTmhRREUzTURjeE9URTBNamN3SGhjTk1qUXdNakEyTURNMU1ESTNXaGNOTXpRd01qQXpNRE0xTURJMwpXakFqTVNFd0h3WURWUVFEREJock0zTXRjMlZ5ZG1WeUxXTmhRREUzTURjeE9URTBNamN3V1RBVEJnY3Foa2pPClBRSUJCZ2dxaGtqT1BRTUJCd05DQUFRV0lySG1uMEFQczQ5bEZKWTdxdFh0R2RRZGV4VGM2Z011VFRRNnZiL2gKbU1yNGVrQzRKamRMMWNhcGxsVkc0aFE1TVFlQ2QrekdMUjIvc01qQ1graVZvMEl3UURBT0JnTlZIUThCQWY4RQpCQU1DQXFRd0R3WURWUjBUQVFIL0JBVXdBd0VCL3pBZEJnTlZIUTRFRmdRVXRoSE9xUm0zRDQvY25rSkpUR1hECm1pUE54S0V3Q2dZSUtvWkl6ajBFQXdJRFNBQXdSUUlnZnp2M2kwVUVmWCtSNDZJK3Uxb1hWeXhjSnRIWGtmOEYKRm1RMFJkWmNsTGtDSVFEV2Y0RkxZWkY1NHhreEdsLzNrc0VaMWtWczIza3p1dm1xYlhHdFl5bDBWQT09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
    server: https://52.202.114.109:6443
  name: default
contexts:
- context:
    cluster: default
    user: default
  name: default
current-context: default
kind: Config
preferences: {}
users:
- name: default
  user:
    client-certificate-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJrVENDQVRlZ0F3SUJBZ0lJQmJ3K1lmWER4R2t3Q2dZSUtvWkl6ajBFQXdJd0l6RWhNQjhHQTFVRUF3d1kKYXpOekxXTnNhV1Z1ZEMxallVQXhOekEzTVRreE5ESTNNQjRYRFRJME1ESXdOakF6TlRBeU4xb1hEVEkxTURJdwpOVEF6TlRBeU4xb3dNREVYTUJVR0ExVUVDaE1PYzNsemRHVnRPbTFoYzNSbGNuTXhGVEFUQmdOVkJBTVRESE41CmMzUmxiVHBoWkcxcGJqQlpNQk1HQnlxR1NNNDlBZ0VHQ0NxR1NNNDlBd0VIQTBJQUJKOGNqZVYwUXE1aTVaU2YKQW50M2ZlTnRYSzQzR1ZWUjB5Ym4ybjUrVlRGWjlRaXN5TEF6bVpIOGNyZTBNWTcwbG9EUHQ1ZHc3eTJrdGhlcwp1YldOUE5talNEQkdNQTRHQTFVZER3RUIvd1FFQXdJRm9EQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0RBakFmCkJnTlZIU01FR0RBV2dCUXpxVlRRNXRGbHFzRDNWRHNGbTFTNnFVVUxJREFLQmdncWhrak9QUVFEQWdOSUFEQkYKQWlFQWxRZ3ppdUxkMEthZnM0cWcySHNnOFNnb2xyakFhTnZWZHdpZ3NGblB4cWNDSUFmSjJzcDhXTnQ3WHUySQp1cTlFMzhOdWJBdzlucndiMW5XR0FpQVpvcm9CCi0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0KLS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJkekNDQVIyZ0F3SUJBZ0lCQURBS0JnZ3Foa2pPUFFRREFqQWpNU0V3SHdZRFZRUUREQmhyTTNNdFkyeHAKWlc1MExXTmhRREUzTURjeE9URTBNamN3SGhjTk1qUXdNakEyTURNMU1ESTNXaGNOTXpRd01qQXpNRE0xTURJMwpXakFqTVNFd0h3WURWUVFEREJock0zTXRZMnhwWlc1MExXTmhRREUzTURjeE9URTBNamN3V1RBVEJnY3Foa2pPClBRSUJCZ2dxaGtqT1BRTUJCd05DQUFRc29ZRXl3bkJOOWFaRzk2WHdJT1ZXU01HUzBWemoyVHBGV3ZabEpSUDgKTjljeGJrYVIrUC9LaWwybW4vcTd4NTZ2KzdUMzVqV2hEVXE2VGlVSCtHWXJvMEl3UURBT0JnTlZIUThCQWY4RQpCQU1DQXFRd0R3WURWUjBUQVFIL0JBVXdBd0VCL3pBZEJnTlZIUTRFRmdRVU02bFUwT2JSWmFyQTkxUTdCWnRVCnVxbEZDeUF3Q2dZSUtvWkl6ajBFQXdJRFNBQXdSUUloQU1DR2M1dVAzYXFYdkNrK3d1VnBpYmYyNnJzQWFCbE0KMHB0YWZVY0x5N0U4QWlBazdqVjhLamx0ZWhkNXFjSm0zdFlCVEV3QzdrTFBBMXZmSXlEeVlLdlBTQT09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
    client-key-data: LS0tLS1CRUdJTiBFQyBQUklWQVRFIEtFWS0tLS0tCk1IY0NBUUVFSUdpeHFtVmx0azdEdnJBRFlqclhOOGxSeE9jeDBVcHNzb2FwUjhzVGxoWGhvQW9HQ0NxR1NNNDkKQXdFSG9VUURRZ0FFbnh5TjVYUkNybUxsbEo4Q2UzZDk0MjFjcmpjWlZWSFRKdWZhZm41Vk1WbjFDS3pJc0RPWgprZnh5dDdReGp2U1dnTSszbDNEdkxhUzJGNnk1dFkwODJRPT0KLS0tLS1FTkQgRUMgUFJJVkFURSBLRVktLS0tLQo=
```
to see if you can find anything, if not, I suggest not to dig into this and wait until the [PR](https://github.com/longhorn/longhorn-tests/pull/1539) is refined and merged. Sorry for the inconvenience.

**Comment by yangchiu** [comment]
Could be difficult to reproduce. The test case could fail due to different reasons instead of what described in https://github.com/longhorn/longhorn/issues/7854#issue-2120056364.

It seems hard to predict when will a backup creation can be triggered by a recurring job after a node reboot.

@james-munson Please don't try to reproduce it by running the test case since it could be diverse. The test case needed to be refined first. Please look into the support bundle or ping me to get the kubeconfig to access the cluster to see if you can find anything, if not, I suggest not to dig into this and wait until the [PR](https://github.com/longhorn/longhorn-tests/pull/1539) is refined and merged. Sorry for the inconvenience.

**Comment by james-munson** [comment]
The error comes from backupstore https://github.com/longhorn/backupstore/blob/9c90a8728ca2e1f95f10a4c6cf67cf7ac20c23e4/lock.go#L104-111

**Comment by james-munson** [comment]
There are 16 such collisions between a DELETION_LOCK and a BACKUP_LOCK for the two volumes over the previous 90 minutes.   So either the deletion is taking surprisingly long, or the recurrence schedule of once a minute is pushing too hard.

**Comment by yangchiu** [comment]
I think it's fair that lock contention causes backup creation and deletion failed. If it makes sense to you as well, feel free to close this issue! cc @innobead

**Comment by james-munson** [comment]
Ok. I think I'll compare it to the issues seen in #7744 first.

**Comment by innobead** [comment]
Removing backport labels first, as this could be related to #7744.

**Comment by james-munson** [comment]
This is not the same as #7744.  The symptom of lock collision is the same, but that one was caused both by a race in the test script itself, and another race in the cleanup of the backupstore when the setting itself was cleared.

This one is just a simple race between a new backup and the retention purge from the last backup such a short time before.  In no case was the existing DELETION_LOCK older than 15 seconds at the time of collision.

I think this kind of collision was anticipated, looking at https://github.com/longhorn/backupstore/blob/b375644f1c5759cf0e8e91263ccf7dda09516004/lock.go#L104

Bottom line is, I think this is behaving as designed.  Pushing a backup and a purge once a minute may be a little aggressive.  I will close this issue, although maybe the test can be slightly modified to recur less often.

**Comment by yangchiu** [comment]
> Bottom line is, I think this is behaving as designed. Pushing a backup and a purge once a minute may be a little aggressive. I will close this issue, although maybe the test can be slightly modified to recur less often.

The test case will be refined in https://github.com/longhorn/longhorn-tests/pull/1539
